README
                                      README for ProBIND

Installation
------------

Please see INSTALL for instructions


What is ProBIND?
----------------

ProBIND is a web application intended for use in the administration of one or
more servers running ISC BIND.

ProBIND is written in the PHP scripting language and uses MySQL to store its
data. The contents of the database are generated into files formatted for ISC
BIND and are pushed to the controlled servers on request through the interface
or by a scheduled job.


Advantages of ProBIND
---------------------

- ProBIND allows for quick searching across domains, allowing for quick editing
  of important information.
  
- ProBIND has built-in templating to allow the use of certain standard
  information across all zones.
  
- Based on settings in ProBIND, NS records will be automatically created.

- ProBIND can automatically populate reverse zones (in-addr.arpa) for all your
  A records. All you need to do is add the zone.


Security Notice
---------------

Because of the high level of access that is currently needed for the operation
of this application, it is highly recommended that you run ProBIND on a non-
public web server. It is further recommended that you set your web server to
use authentication for access to the application. 


Limitations
-----------

- The import script is rather limited and doesn't work well with extended
  features (GENERATE is an example) or odd formatting.

- The import script cannot handle zone files larger than 64 kilobytes

- Currently, a server can be only a master or a slave. Mixing of roles is not
  possible.
  
- ProBIND can only automatically handle in-addr.arpa zones for /24 networks.
  Other sizes can only be handled manually at this time. 

- Only a subset of all possible record types is supported. These include A, CNAME,
  MX, NS, PTR, TXT, and HINFO.

- ProBIND exists independently of ISC BIND, thus no feedback from BIND itself
  is used. This may change in the future, although it will be limited due to
  the nature of ProBIND.


Acknowledgements
----------------

Current maintainer: Michael Johnson <youngmug@users.sourceforge.net>

I would like to acknowledge PhD Computing for offering support and time to
contribute to this project and get it started again.


ProBIND 2 Maintainer: Alexei P. Roudnev


Original Author: Flemming S. Johansen <fjohansen@proventum.net>

I acknowledge the role of my colleague Lasse L. Johnsen, who has
contributed a lot of useful suggestions, without which this software
would almost certainly be utterly useless to anyone other than myself.


Logo: Susanne B. Olesen


Visual Layout: Jeffrey R. McDonald


History
--------------------

ProBIND was originally developed by Flemming S. Johansen as part of his duties
as resident DNS manager at Proventum Solutions (now Consendo Systems A/S).

Later, a fork of ProBIND called ProBIND2 was developed by Alexei P. Roudnev, a
senior network/software engineer, at Exigen Group LTD.

ProBIND2 was re-incorporated in ProBIND by the new maintainer, Michael Johnson,
a systems administrator for PhD Computing,


License
-------

This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2 of the License, or (at your
option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.



Last Update: 20100615

Changelog
ProBIND 1.0.4 (?????)
---------------------

- Bug 1981649 - Updated root.hints
- Bug 2796935 - Zone options should be generated for slave server.
- Request 1380317 - Added IPv6 AAAA record support
- Request 1928824 - Added basic chroot support
- Multiple Bugs - Import script cleanups
- Adjusted root.hints so EOL isn't stripped


ProBIND 1.0.3 (10 November 2008)
--------------------------------

It's been some time since the last release, and I was hoping to have more
done. Unfortunately, some big projects came up, and re-working things like the
import system will take some time, so it's best to make a release with 
bugfixes completed.

- Bug 1885891 - Deletion of zone broken
- Bug 1981649 - Add updated root.hints
- Bug 1928820 - Clarify reverse zone population in documentation
- Clean code to load config.php in lib.inc
- Clean code to load libraries in import script
- Remove broken "help" links that go to manual.html


ProBIND 1.0.2 (25 December 2007)
--------------------------------

After a long hiatus, the ProBIND project is alive again! Welcome to those
updating to this release and to those new to the software. This release
incorporates the work done in the ProBIND2 fork (now defunct) and involves
cleanup work to allow ProBIND to function on modern configurations. There is
otherwise no major functionality changes in this release. Please look over the
TODO file for future plans.

- Merged in changes from ProBIND2.
- Cleaned up unused files from merge.
- Removal of ProBIND2 namespace system.
- Removed forced auth on update - webserver should be configured to
  handle, not us.
- Changed default remote user to root in the push scripts.
- Miscelaneous cleanup involving replacing short open tags with full tags
  and clean up uninitialized variables.
- Updates to INSTALL, README, TODO files
- Added upgrade instructions in UPGRADE


ProBIND2 Version 2.0 build 1
=====================
Bug fixes in:
- import
- reverse zone generations
- special cases when zone is empty etc
- scripts - bug fixed for local script
- serial generation
- many small bugs fixed.

Version 2.1 - TODO list prepared.
Alexei Roudnev


ProBIND2 Version 2.0
===============
Released April 2003

Changes:
- new zone editor
- new update alghoritm
- zone comments
- PTR flag
- new update programms
- support for named.conf options
- support for zone options
- embedded RR testing
- 'update' block rewritten 
- now slaves can be configured to keep secondary zones
- update scripts can 'cvs' files
- when creating subzone, records moved into it if theey belong to subzone
- additional rights for 'update' required
- numeroous bug fixes.
- new documentation
- new 'import'
- system deployed in production, production internal, office, office internal environment
Alexei Roudnev



Version 1.0.1
=============

Released 15-feb-2001, mostly a bugfix release

- Updated the INSTALL document with some badly needed troubleshooting tips.
- Updated the TODO list with a few new ideas
- Added the tools/full-report.php and bin/domain-report scripts, updated 
  tools/toolmenu and manual.html to match.
- Added the sbin/copy-local update script
- Fixed a bug in tools/ip-ranges which could cause incomplete lists to
  be displayed.
- Fixed a missing </BODY></HTML> in addzone.php and update.php
- Fixed some marginal SQL SELECT syntax in several places
- Fixed a bad mysql call in inc/lib.php
- Fixed a bad library call in addzone.php
- Changed addzone.php to allow bulk zone creation
- Made bin/mknamed.conf work in the absence of slave servers
- Fixed bin/mkzonefile to add a missing TTL in zone files
- Added an IP number check to inc/lib.inc:validate_record

Version 1.0
===========

Released 19-dec-2000


brrec.php
<?php
require 'inc/lib.inc';

$ROWS_PER_PAGE = 50;

$html_top = "<html><body>";


$start_frame = '
<HTML>
<HEAD>
<TITLE>Browse Records</TITLE>
<LINK rel="stylesheet" href="style.css" type="text/css">
</HEAD>
<FRAMESET rows="12,*" frameborder="0" border="0" framespacing="0">
  <FRAME src="topshadow.html" name="topshadow" noresize scrolling=no frameborder ="0" border="0" framespacing="0" marginheight="0" marginwidth="0">
  <FRAME src="brrec.php?frame=brrec" name="main" noresize scrolling=auto frameborder="0" border="0" framespacing="0" marginheight="0" marginwidth="10">
</FRAMESET>
</HTML>
';

$query_form = '
<HTML><HEAD>
<TITLE>Resource Record Query Form</TITLE>
<LINK rel="stylesheet" href="style.css" type="text/css">
</HEAD><BODY bgcolor="cccc99" background="images/BG-shadowleft.gif">
<FORM action="brrec.php" method="post">
<INPUT type=hidden name="select" value="result">
<INPUT type=hidden name="pointer" value="%d">
<TABLE width="100%%">
<TR><TH>'.trans('Zone').'</TH><TH>'.trans('Domain').'</TH><TH>'.trans('Type').'</TH><TH>'.trans('Pref').'</TH><TH>'.trans('Data').'</TH></TR>
<TR>
<TD><INPUT TYPE=text name="zdomain" value="%s" SIZE=15 MAXLENGTH=100 onfocus=this.select()></TD>
<TD><INPUT TYPE=text name="rdomain" value="%s" SIZE=15 MAXLENGTH=100 onfocus=this.select()></TD>
%s
<TD><INPUT type=text name="pref" value="%s" SIZE=2 MAXLENGTH=4 onfocus=this.select()></TD>
<TD><INPUT type=text name="data" value="%s" SIZE=25></TD>
<TD><INPUT type=submit name="submit" value="'.trans('Search').'"></TD>
<TD align=right><A HREF="manual.html#records">Help</A></TD>
</TR></TABLE>
</FORM>
Use the %% wildcard character to specify a substring search.
<HR>
';

$query_type = '
<TD><SELECT name="type">
	<OPTION>*</OPTION>
	<OPTION>A</OPTION>
	<OPTION>CNAME</OPTION>
	<OPTION>MX</OPTION>
	<OPTION>NS</OPTION>
	<OPTION>PTR</OPTION>
	<OPTION>TXT</OPTION>
	<OPTION>HINFO</OPTION>
	<OPTION>SRV</OPTION>
</SELECT></TD>
';

$empty_result = '
<HR><BR>
<H1>Resource Record Query Results</H1>
<BR><HR><BR>
';

$button_form = '
<FORM action="brrec.php" method="post">
<INPUT type=hidden name="select" value="result">
<INPUT type=hidden name="zdomain" value="%s">
<INPUT type=hidden name="rdomain" value="%s">
<INPUT type=hidden name="type" value="%s">
<INPUT type=hidden name="pref" value="%s">
<INPUT type=hidden name="data" value="%s">
<INPUT type=hidden name="pointer" value="%s">
<INPUT type=submit name="%s" value="%s">
</FORM>
';

$html_close = "</BODY></HTML>\n";

function query_type($type)
{
	$result = "<TD><SELECT name=\"type\">\n";
	$result .= sprintf("<OPTION%s>*</OPTION>\n",
		((!$type || ($type == '*')) ? ' selected' : ''));
	$types = array('A', 'AAAA', 'CNAME', 'MX', 'NS', 'PTR', 'TXT', 'HINFO', 'SRV');
	while ($tp = each($types)) {
		$result .= sprintf("<OPTION%s>%s</OPTION>\n",
			($type == $tp[1] ? ' selected' : ''),
			$tp[1]);
	}
	$result .= "</SELECT></TD>\n";
	return $result;
}

function query_form($INPUT_VARS)
{
	global $query_form;

	$zdomain = '';
	$rdomain = '';
	$type = '';
	$pref = '';
	$data = '';
	$pointer = '';

	//Check if the form has been submitted before
	if(isset($INPUT_VARS['zdomain']))
	{
		$zdomain = ltrim(rtrim($INPUT_VARS['zdomain']));
		$rdomain = ltrim(rtrim($INPUT_VARS['rdomain']));
		$type = ltrim(rtrim($INPUT_VARS['type']));
		$pref = ltrim(rtrim($INPUT_VARS['pref']));
		$data = ltrim(rtrim($INPUT_VARS['data']));
		$pointer = ltrim(rtrim($INPUT_VARS['pointer']));
	}
	return sprintf($query_form,
		$pointer, $zdomain, $rdomain, query_type($type), $pref, $data);
}

function result_form($INPUT_VARS)
{
	global $empty_result, $button_form, $ROWS_PER_PAGE;
	$result = "";
	$zdomain = ltrim(rtrim($INPUT_VARS['zdomain']));
	$rdomain = ltrim(rtrim($INPUT_VARS['rdomain']));
	$type = ltrim(rtrim($INPUT_VARS['type']));
	$pref = ltrim(rtrim($INPUT_VARS['pref']));
	$data = ltrim(rtrim($INPUT_VARS['data']));
	$pointer = ltrim(rtrim($INPUT_VARS['pointer']));
	if (!strlen($zdomain.$rdomain.$pref.$data) && $type == '*')
		return $empty_result;
	$queryhead = "SELECT zones.id AS zid, zones.domain AS zdom, records.domain AS rdom, records.id AS rrid, type, pref, data FROM zones, records ";
	$counthead = "SELECT count(*) FROM zones, records ";
	$querycond = "WHERE zones.id = records.zone AND zones.domain != 'TEMPLATE' AND type != 'SOA'".access();
	if (strlen($zdomain)) {
		if (strchr($zdomain, "%"))
			$querycond .= " AND zones.domain like '$zdomain'";
		else
			$querycond .= " AND zones.domain = '$zdomain'";
	}
	if (strlen($rdomain)) {
		if (strchr($rdomain, "%"))
			$querycond .= " AND records.domain like '$rdomain'";
		else
			$querycond .= " AND records.domain = '$rdomain'";
	}
	if ($type != '*')
		$querycond .= " AND type = '$type'";
	if (strlen($pref))
		$querycond .= " AND pref = $pref";
	if (strlen($data)) {
		if (strchr($data, "%"))
			$querycond .= " AND records.data like '$data'";
		else
			$querycond .= " AND records.data = '$data'";
	}
	$querytail = " ORDER BY zones.domain, records.domain";
	$querytail .= " LIMIT $pointer, ".($pointer+$ROWS_PER_PAGE);
	$rid = sql_query($counthead.$querycond);
	$row = mysql_fetch_row($rid);
	$count = $row[0];
	mysql_free_result($rid);
	$displ = ( (($pointer + $ROWS_PER_PAGE) <= $count) ?
		$ROWS_PER_PAGE : ($count % $ROWS_PER_PAGE) );
	$result .= "Displaying $displ of $count records, starting at $pointer<BR>\n";
	$rid = sql_query($queryhead.$querycond.$querytail);
#	$result .= "Found ".($count - $pointer)." matching records.<BR>\n";
	$result .= "<TABLE><TR>\n";
	$result .= "<TD>";
	$result .= sprintf($button_form,
		$zdomain, $rdomain, $type, $pref, $data,
		0, "start",
		"<<<< ".trans("First"));
	$result .= "</TD>";
	if ($pointer) {
		$result .= "<TD>";
		$result .= sprintf($button_form,
			$zdomain, $rdomain, $type, $pref, $data,
			$pointer-$ROWS_PER_PAGE, "prev",
			"<< ".trans("Prev")." $ROWS_PER_PAGE");
		$result .= "</TD>";
	}
	if (($count >= 50) && (($pointer + $ROWS_PER_PAGE) <= $count)) {
		$result .= "<TD>";
		$result .= sprintf($button_form,
			$zdomain, $rdomain, $type, $pref, $data,
			$pointer+$ROWS_PER_PAGE, "next",
			trans("Next")." $ROWS_PER_PAGE >>");
		$result .= "</TD>";
	}
	$result .= "<TD>";
	$result .= sprintf($button_form,
		$zdomain, $rdomain, $type, $pref, $data,
		($count - ($count % $ROWS_PER_PAGE)), "last",
		trans("Last")." >>>>");
	$result .= "</TD>";
	$result .= "</TR></TABLE>\n";
	$result .= "<TABLE><TR><TH align=left>".trans('Host')."</TH><TH align=left>".trans('Domain')."</TH><TH align=left>".trans('Type')."</TH><TH align=left>".trans('Pref')."</TH><TH align=left>".trans('Data')."</TH></TR>\n";
	while ($row = mysql_fetch_array($rid)) {
		$result .= "<TR><TD>".decode($row['rdom'])."</TD>";
		$result .= "<TD><A HREF=\"brzones.php?frame=records&zone=";
		$result .= $row['zid']."&rrid=".$row['rrid']."\">".decode($row['zdom'])."</A></TD>";
		$result .= "<TD>".$row['type']."</TD>";
		$result .= "<TD>".$row['pref']."</TD>";
		$result .= "<TD>".decode($row['data'])."</TD></TR>\n";
	}
	$result .= "</TABLE>\n";
	mysql_free_result($rid);
	return $result;
}

#
# MAIN
#

get_input();

$selectval = isset($INPUT_VARS['select'])? $INPUT_VARS['select'] : ''; 
if (!$selectval) {
	$selectval='result';
	$INPUT_VARS['type']='*';
	$INPUT_VARS['zdomain']='%';
	$INPUT_VARS['rdomain']='%';
	$INPUT_VARS['pref']='';
	$INPUT_VARS['data']='';
	$INPUT_VARS['pointer']='0';
}

switch($selectval) {
case 'result':
	print query_form($INPUT_VARS);
	print result_form($INPUT_VARS);
	break;
case 'query':
	print query_form($INPUT_VARS);
	break;
default:
	if (isset($INPUT_VARS['frame']) && $INPUT_VARS['frame'] == 'brrec')
		print $html_top . query_form(array());
	else
		print $start_frame;
}
print $html_close;
?>
